streaming/streaming.go

```diff
@@ -635,24 +642,40 @@ func LaunchDataflowJob(ctx context.Context, targetProfile profiles.TargetProfile
 		inputFilePattern = inputFilePattern + "/"
 	}
 	fmt.Println("Reading files from datastream destination ", inputFilePattern)
-	var dataflowHostProjectId string
-	if dataflowCfg.HostProjectId == "" {
-		dataflowHostProjectId, _ = utils.GetProject()
-	} else {
-		dataflowHostProjectId = dataflowCfg.HostProjectId
-	}
-
-	dataflowSubnetwork := ""
 
-	// If custom network is not selected, use public IP. Typical for internal testing flow.
-	workerIpAddressConfig := dataflowpb.WorkerIPAddressConfiguration_WORKER_IP_PUBLIC
-
-	if dataflowCfg.Network != "" {
+	// Initiate runtime environment flags and overrides.
+	var (
+		dataflowProjectId        = project
+		dataflowVpcHostProjectId = project
+		gcsTemplatePath          = DEFAULT_TEMPLATE_PATH
+		dataflowSubnetwork       = ""
+		workerIpAddressConfig    = dataflowpb.WorkerIPAddressConfiguration_WORKER_IP_PUBLIC
+		dataflowUserLabels       = make(map[string]string)
+	)
+	// If project override present, use that otherwise default to Spanner project. Useful when customers want to run Dataflow in separate project.
+	if dataflowCfg.ProjectId != "" {
+		dataflowProjectId = dataflowCfg.ProjectId
+	}
+	// If VPC Host project override present, use that otherwise default to Spanner project.
+	if dataflowCfg.VpcHostProjectId != "" {
+		dataflowVpcHostProjectId = dataflowCfg.VpcHostProjectId
+	}
+	if dataflowCfg.GcsTemplatePath != "" {
+		gcsTemplatePath = dataflowCfg.GcsTemplatePath
+	}
+
+	// If either network or subnetwork is specified, set IpConfig to private.
+	if dataflowCfg.Network != "" || dataflowCfg.Subnetwork != "" {
 		workerIpAddressConfig = dataflowpb.WorkerIPAddressConfiguration_WORKER_IP_PRIVATE
-		if dataflowCfg.Subnetwork == "" {
-			return internal.DataflowOutput{}, fmt.Errorf("if network is specified, subnetwork cannot be empty")
-		} else {
-			dataflowSubnetwork = fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/subnetworks/%s", dataflowHostProjectId, dataflowCfg.Location, dataflowCfg.Subnetwork)
+		if dataflowCfg.Subnetwork != "" {
+			dataflowSubnetwork = fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/subnetworks/%s", dataflowVpcHostProjectId, dataflowCfg.Location, dataflowCfg.Subnetwork)
+		}
+	}
+
+	if dataflowCfg.AdditionalUserLabels != "" {
+		err = json.Unmarshal([]byte(dataflowCfg.AdditionalUserLabels), &dataflowUserLabels)
+		if err != nil {
+			return internal.DataflowOutput{}, fmt.Errorf("could not unmarshal AdditionalUserLabels json %s : error = %v", dataflowCfg.AdditionalUserLabels, err)
```

Comment:
Should we fail on improperly configured User labels or should this be a warning?

Reply:
Thats a good question, normally I would keep a warning, but dataflow does not allow the user to add a label once the job has started (unlike other resources). They would either miss it and the job would run without a label or they notice labels werent applied, cancel the job and manually retrigger. 

I chose to fail but open to alternatives.

Reply:
With the new regex check, we wont need this anymore